gdnative-core/src/dictionary.rs

```diff
@@ -11,121 +12,164 @@ use crate::Variant;
-        unsafe { Variant((get_api().godot_dictionary_get)(&self.0, &key.0)) }
+        unsafe { Variant((get_api().godot_dictionary_get)(self.sys(), key.sys())) }
     }
 
-    /// Sets a value to the element corresponding to the key.
+    /// Update an existing element corresponding ot the key.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the entry for `key` does not exist.
     #[inline]
-    pub fn set(&mut self, key: &Variant, val: &Variant) {
-        unsafe { (get_api().godot_dictionary_set)(&mut self.0, &key.0, &val.0) }
+    pub fn update(&self, key: &Variant, val: &Variant) {
+        assert!(self.contains(key), "Can only update entries that exist");
+        unsafe { (get_api().godot_dictionary_set)(self.sys_mut(), key.sys(), val.sys()) }
     }
 
     /// Returns a reference to the value corresponding to the key.
+    ///
+    /// # Safety
+    ///
+    /// The returned reference is invalidated if the same container is mutated through another
+    /// reference.
+    ///
+    /// `Variant` is reference-counted and thus cheaply cloned. Consider using `get` instead.
     #[inline]
-    pub fn get_ref(&self, key: &Variant) -> &Variant {
-        unsafe {
-            Variant::cast_ref((get_api().godot_dictionary_operator_index_const)(
-                &self.0, &key.0,
-            ))
-        }
+    pub unsafe fn get_ref(&self, key: &Variant) -> &Variant {
+        Variant::cast_ref((get_api().godot_dictionary_operator_index_const)(
+            self.sys(),
+            key.sys(),
+        ))
     }
 
     /// Returns a mutable reference to the value corresponding to the key.
+    ///
+    /// # Safety
+    ///
+    /// The returned reference is invalidated if the same container is mutated through another
+    /// reference. It is possible to create two mutable references to the same memory location
+    /// if the same `key` is provided, causing undefined behavior.
     #[inline]
-    pub fn get_mut_ref(&mut self, key: &Variant) -> &mut Variant {
-        unsafe {
-            Variant::cast_mut_ref((get_api().godot_dictionary_operator_index)(
-                &mut self.0,
-                &key.0,
-            ))
-        }
+    #[allow(clippy::mut_from_ref)]
+    pub unsafe fn get_mut_ref(&self, key: &Variant) -> &mut Variant {
+        Variant::cast_mut_ref((get_api().godot_dictionary_operator_index)(
+            self.sys_mut(),
+            key.sys(),
+        ))
     }
 
     /// Returns a GodotString of the `Dictionary`.
     #[inline]
     pub fn to_json(&self) -> GodotString {
-        unsafe { GodotString((get_api().godot_dictionary_to_json)(&self.0)) }
+        unsafe { GodotString((get_api().godot_dictionary_to_json)(self.sys())) }
     }
 
     /// Returns an array of the keys in the `Dictionary`.
     #[inline]
-    pub fn keys(&self) -> VariantArray {
-        unsafe { VariantArray((get_api().godot_dictionary_keys)(&self.0)) }
+    pub fn keys(&self) -> VariantArray<Shared> {
```

Comment:
It seems that `godot_dictionary_keys` should return a new array, but I'm not sure if this is something we should depend on: https://github.com/godotengine/godot/blob/master/core/dictionary.cpp#L193

Reply:
I don't understand what the issue is there. This function always returns an `Array` in all code paths, unless I am missing something?

Reply:
I was implying that the return value could be `VariantArray<Unique>`, because the array is newly created in the C++ code, if we feel that this is something we can depend on.

Reply:
This code didn't change in a long time I think, so I would say it's safe to depend on that, although this might have to be broken in future while a `Shared` array would always be correct.

I don't mind either way, it probably makes sense to make it `Unique` actually.

Reply:
I think it's ok to make it `Unique` here for now. It doesn't make sense for the array to *not* be new, because then it can be modified and made incorrect.

Reply:
It doesn't make much sense from an API perspective that `keys()` or `values()` would return a shared array, so I made both return `VariantArray<Unique>` now.